### PR TITLE
REL: 5.0.0

### PR DESCRIPTION
**Pull request by pubpub-zz**
## Version 5.0.0, 2024-09-15

This release drops support to Python 3.7 and introduces the capability to build incremental PDF files.

### Security (SEC)
- Fix GitHub workflow vulnerable to script injection (#2787)

### New Features (ENH)
- Add capability to remove /Info from PDF (#2820)
- Robustify parsing for Object streams in XRef rebuilding (#2818)
- Add incremental capability to PdfWriter (#2811)
- Add UniGB-UTF16 encodings (#2819)
- Robustify on missing font for Tf  operator in text_extract() (#2816)
- Accept utf strings for metadata (#2802)
- Report PdfReadError instead of RecursionError (#2800)
- Compress PDF files merging identical objects (#2795)

### Bug Fixes (BUG)
- Warn when visitor* arguments are ignored (#2845)
- Fix sheared image (#2801)
- Handle Sequence as an IndirectObject when extracting text with layout mode (#2788)

### Robustness (ROB)
- Robustify .set_data() (#2821)
- Raise PdfReadError when missing /Root in trailer (#2808)
- Fix extract_text() issues on damaged PDFs (#2760)
- Handle images with empty data when processing an image from bytes (#2786)

### Documentation (DOC)
- Small changes to PaperSize notes (#2834)

### Developer Experience (DEV)
- Fix coverage uploads (#2832)
- Test against Python 3.13 (#2776)

### Maintenance (MAINT)
- Deprecate PdfMerger, AnnotationBuilder and other deprecations cleanup (#2813)
- simplify test with None and NullObject
- Improve wording of set_data error message (#2810)
- Remove unused code (#2805)
- Fix mypy type output (#2799)
- Improve test coverage (#2796)
- Remove b_ and str_ (#2792)
- Drop Python 3.7 support (#2793)
- Remove unused AnnotationFlag
- Remove unused paeth_predictor (#2773)

### Code Style (STY)
- Use f-string = functionality (#2835)
- Refactor b_ (#2772)
- Remove boolean value comparison (#2779)
- Increase readability (#2775)
- Reorder exported symbols (#2774)
- Amend B_CACHE key types (#2766)

### Other

[Full Changelog](https://github.com/py-pdf/pypdf/compare/4.3.1...5.0.0)

### CODE REVIEW

**Review by stefan6419846 — DISMISSED** [review]
Please reduce the changelog to omit unimportant changes and fix line 15 to match the other lines. (Please note that this has to be done for the merge commit data as well.) Additionally, I slightly adjusted the wording of the introductory text which should be transferred as well.

If you need help, feel free to come back to me and I will have a look at condensing the changelog myself tomorrow.

**Comment by MartinThoma** [comment]
@pubpub-zz @stefan6419846 Thank you for keeping pypdf alive :pray:

File: CHANGELOG.md
Comment (MartinThoma) [inline CHANGELOG.md]:
```suggestion
```

That exact string is mentioned a few lines before

File: CHANGELOG.md
Comment (MartinThoma) [inline CHANGELOG.md]:
```suggestion
```

That exact string is mentioned a few lines before

**Comment by pubpub-zz** [comment]
@stefan6419846 
Is it ok now ?

**Comment by MartinThoma** [comment]
Python 3.7 support was removed in https://github.com/py-pdf/pypdf/pull/2793

That should be mentioned in the Changelog as a breaking change:

```
### Deprecations (DEP)
-  Drop Python 3.7 support (#2793)
```

File: CHANGELOG.md
Comment (MartinThoma) [inline CHANGELOG.md:5]:
```suggestion
### Deprecations (DEP)
```

**Review by MartinThoma — APPROVED** [review]
Looks good to me :tada:  Thank you :pray: 

As I'm out of touch of recent developments it would be good to wait for the approval of @stefan6419846

File: CHANGELOG.md
Comment (stefan6419846) [inline CHANGELOG.md:3]: Could we please have the introductory text with the highlights here as well?

File: Makefile
Comment (stefan6419846) [inline Makefile:1]: Please do not change this file in this PR - this is unrelated and unnecessary.

File: pyproject.toml
Comment (stefan6419846) [inline pyproject.toml:33]: This is wrong - the whole line can be removed.

File: pyproject.toml
Comment (stefan6419846) [inline pyproject.toml:44]: Please do these cleanups in a separate PR.

File: pyproject.toml
Comment (stefan6419846) [reply pyproject.toml:33]: Please revert and do such cleanups in a different PR.

File: pyproject.toml
Comment (stefan6419846) [inline pyproject.toml:49]: This is wrong and should be discussed in another PR.

File: pyproject.toml
Comment (stefan6419846) [inline pyproject.toml:51]: There is no need to update this pin.

File: CHANGELOG.md
Comment (pubpub-zz) [reply CHANGELOG.md:6]: They were just deprecating, providing a warning. Now they have been removed as it was announced

File: pyproject.toml
Comment (pubpub-zz) [reply pyproject.toml:51]: There is an issue on previous version

Comment (stefan6419846) [reply pyproject.toml:51]: This is wrong if I remember correctly - there is one specific version which does not work. If we would like to avoid users installing it, we could declare it as a condition, but this condition is wrong in my opinion, especially introducing this in the release commit without further discussions.

File: CHANGELOG.md
Comment (stefan6419846) [reply CHANGELOG.md:6]: We might want to adapt this in the changelog nevertheless, for example using something like "Remove the deprecated PfdMerger and AnnotationBuilder classes and other deprecations cleanup"

File: Makefile
Comment (pubpub-zz) [reply Makefile:1]: this makefile refers to python 3.7 : I've updated it to match drop any reference to this release

File: pyproject.toml
Comment (pubpub-zz) [reply pyproject.toml:44]: the current code of pypdf is no more compatible with python 3.7.
I've just remove any reference to this release to prevent mistakes

File: pyproject.toml
Comment (pubpub-zz) [reply pyproject.toml:49]: Ok to discuss it later

File: pyproject.toml
Comment (stefan6419846) [reply pyproject.toml:44]: We did not do this before, although we might have introduced incompatibilities. Apart from this, this is still unrelated to the release PR. If you want to change this, please propose this in a separate PR.

File: pyproject.toml
Comment (pubpub-zz) [reply pyproject.toml:33]: can you please clarify. Unless I'm wrong, we still have `from dataclasses import dataclass` in _pages.py and _utils.py

File: CHANGELOG.md
Comment (pubpub-zz) [reply CHANGELOG.md:3]: added

File: pyproject.toml
Comment (MasterOdin) [reply pyproject.toml:33]: `dataclasses` was introduced in Python 3.7 via [PEP-557](https://peps.python.org/pep-0557/), and so can rely on the [builtin module](https://docs.python.org/3/library/dataclasses.html) for 3.7.

File: pyproject.toml
Comment (pubpub-zz) [reply pyproject.toml]: I was aware that both would be installed. I wanted to still show compatbility with both libraries although cryptography is recommanded. I revert and let you propose a PR for this

File: CHANGELOG.md
Comment (pubpub-zz) [reply CHANGELOG.md:9]: The wording is in line with the previous entries in changelog.md. However OK for this rewording

File: pyproject.toml
Comment (pubpub-zz) [reply pyproject.toml:44]: I personnally see no advantage to handle that in a separate PR. as I've asked you to manage this clean up whilst managing the crypto section

File: pyproject.toml
Comment (MasterOdin) [reply pyproject.toml:50]: I had started on a PR to remove `PyCryptodome` altogether (including the file that references it), which I can look to finish up in next week or so.

File: CHANGELOG.md
Comment (MasterOdin) [reply CHANGELOG.md]: I agree with @stefan6419846's rewording as I think it's more correct in its intention which is that the code was removed and it no longer usable. You _have_ to move off PdfReader. For `PdfMerger` specifically, for v3.9.1, there was already:

> * Deprecate PdfMerger (#1866)

So it doesn't make sense to me that it's been doubly deprecated? The use of "Remove" is inline with the v4 wording as well:

> * Remove deprecated code (#2367)
> * Remove deprecated XMP properties (#2386)

File: pyproject.toml
Comment (MartinThoma) [inline pyproject.toml:49]: cryptography still works with  python 3.7, right? Just because we no longer support 3.7, it doesn't mean that we have to make it actively harder to use pypdf with 3.7.

Comment (pubpub-zz) [reply pyproject.toml:49]: the specific code for python 3.7 has been removed.  pypdf should not work anymore on this release

File: pyproject.toml
Comment (MartinThoma) [reply pyproject.toml:52]: If there is a new feature in 10.0 which we require then `>=10.0` is the way to go.

If we know that it works from 10.0 upwards, but e.g. 10.2 was broken, we can also do something like `>=10.0,!=10.2`. See https://peps.python.org/pep-0440/#examples

> I would not like to allow issues to be raised because not using the good version.

:+1: 

> as we move to drop python 3.7, pillow needs to be on version <=10.0

I guess you mean >= 10.0?

However, in that case Python will automatically select the right one. We should typically only restrict the version if there are features missing (or an API changed so we know it's incompatible)

> why you want to stay on obsolete version ?

It's not that. A library should be as permissive as possible with the version as the application using it will likely also use other libraries. The more restrictive we are, the harder it will be to use pypdf in combination with other libraries.

Users can still install a newer version.

File: CHANGELOG.md
Comment (stefan6419846) [inline CHANGELOG.md:5]:
```suggestion
This version drops support for Python 3.7 (not maintained since July 2023), PdfMerger (use PdfWriter instead) and AnnotationBuilder (use annotations instead).
```

**Comment by stefan6419846** [comment]
I have just reverted all the unrelated changes to the files. They need further dedicated discussions.

**Review by stefan6419846 — APPROVED** [review]
Should be okay now. Please ensure that the merge commit uses the correct (cleaned up) message. Creating the tag will probably need manual tasks as well.

**Pull request merged**
